src/mapnode.h

```diff
@@ -187,53 +207,97 @@ struct MapNode
-
-	void setLight(LightBank bank, u8 a_light, const NodeDefManager *nodemgr);
+	void setLight(LightBank bank, u8 a_light, ContentLightingFlags f) noexcept
+	{
+		// If node doesn't contain light data, ignore this
+		if (!f.has_light)
+			return;
+		if(bank == LIGHTBANK_DAY)
+		{
+			param1 &= 0xf0;
+			param1 |= a_light & 0x0f;
+		}
+		else if(bank == LIGHTBANK_NIGHT)
+		{
+			param1 &= 0x0f;
+			param1 |= (a_light & 0x0f)<<4;
+		}
+		else
+			assert("Invalid light bank" == NULL);
+	}
 
 	/**
 	 * Check if the light value for night differs from the light value for day.
 	 *
 	 * @return If the light values are equal, returns true; otherwise false
 	 */
-	bool isLightDayNightEq(const NodeDefManager *nodemgr) const;
+	bool isLightDayNightEq(ContentLightingFlags f) const
+	{
+		bool isEqual;
+
+		if (f.has_light) {
+			u8 day   = MYMAX(f.light_source, param1 & 0x0f);
+			u8 night = MYMAX(f.light_source, (param1 >> 4) & 0x0f);
+			isEqual = day == night;
+		} else {
+			isEqual = true;
+		}
 
-	u8 getLight(LightBank bank, const NodeDefManager *nodemgr) const;
+		return isEqual;
+	}
+
+
+	u8 getLight(LightBank bank, ContentLightingFlags f) const noexcept
+	{
+		// Select the brightest of [light source, propagated light]
+		u8 light;
+		if(f.has_light)
+			light = bank == LIGHTBANK_DAY ? param1 & 0x0f : (param1 >> 4) & 0x0f;
+		else
+			light = 0;
+		return MYMAX(f.light_source, light);
+	}
 
 	/*!
 	 * Returns the node's light level from param1.
 	 * If the node emits light, it is ignored.
-	 * \param f the ContentFeatures of this node.
+	 * \param f the ContentLightingFlags of this node.
 	 */
-	u8 getLightRaw(LightBank bank, const ContentFeatures &f) const noexcept;
+	u8 getLightRaw(LightBank bank, ContentLightingFlags f) const noexcept
+	{
+		if(f.has_light)
+			return bank == LIGHTBANK_DAY ? param1 & 0x0f : (param1 >> 4) & 0x0f;
+		return 0;
+	}
 
-	/**
-	 * This function differs from getLight(LightBank bank, NodeDefManager *nodemgr)
-	 * in that the ContentFeatures of the node in question are not retrieved by
-	 * the function itself.  Thus, if you have already called nodemgr->get() to
-	 * get the ContentFeatures you pass it to this function instead of the
-	 * function getting ContentFeatures itself.  Since NodeDefManager::get()
-	 * is relatively expensive this can lead to significant performance
-	 * improvements in some situations.  Call this function if (and only if)
-	 * you have already retrieved the ContentFeatures by calling
-	 * NodeDefManager::get() for the node you're working with and the
-	 * pre-conditions listed are true.
-	 *
-	 * @pre f != NULL
-	 * @pre f->param_type == CPT_LIGHT
-	 */
-	u8 getLightNoChecks(LightBank bank, const ContentFeatures *f) const noexcept;
+	bool getLightBanks(u8 &lightday, u8 &lightnight, ContentLightingFlags f) const
+	{
+		// Select the brightest of [light source, propagated light]
+		if(f.has_light)
+		{
+			lightday = param1 & 0x0f;
+			lightnight = (param1>>4)&0x0f;
```

Comment:
Wouldn't a bitfield struct be easier to deal with than manual shift operations every time?